lib/libshare/nfs.c

```diff
@@ -323,19 +260,113 @@ add_linux_shareopt(char **plinux_opts, const char *key, const char *value)
 	return (SA_OK);
 }
 
+static int
+update_host_list(void *cookie)
+{
+	char **plinux_opts = (char **)cookie;
+	nfs_share_list_t *cur_share, *new_share;
+
+	/* Get the last entry in the list. */
+	cur_share = list_tail(&all_nfs_shares_list);
+	if (cur_share == NULL)
+		return (SA_OK);
+
+	/* Create a new object list */
+	new_share = (nfs_share_list_t *)malloc(sizeof (nfs_share_list_t));
+	if (new_share == NULL)
+		return (SA_NO_MEMORY);
+	list_link_init(&new_share->next);
+
+	/* If the current list is empty, the new list is 'world'. */
+	if (cur_share == NULL)
+		strcpy(new_share->host, "*");
+	else
+		strcpy(new_share->host, cur_share->host);
+	sprintf(new_share->opts, "%s", *plinux_opts);
+
+	/* Replace the old head with this new object */
+	if (cur_share->host != NULL)
+		list_link_replace(&cur_share->next, &new_share->next);
+	else
+		list_insert_tail(&all_nfs_shares_list, new_share);
+
+	return (SA_OK);
+}
+
 /*
  * Validates and converts a single Solaris share option to its Linux
- * equivalent.
+ * equivalent, verifies that the option is valid and then stores it
+ * in "cookie"..
  */
 static int
 get_linux_shareopts_cb(const char *key, const char *value, void *cookie)
 {
 	char **plinux_opts = (char **)cookie;
+	nfs_share_list_t *opts = NULL;
 
-	/* host-specific options, these are taken care of elsewhere */
-	if (strcmp(key, "ro") == 0 || strcmp(key, "rw") == 0 ||
-	    strcmp(key, "sec") == 0)
-		return (SA_OK);
+// TODO: Make sure that a 'sec=yyy' (or any other option) BEFORE a/any
+//  'rw[=@xxx]' or 'ro[=@xxx]' works.
+	if (strcmp(key, "ro") == 0 || strcmp(key, "rw") == 0) {
+
+		opts = (nfs_share_list_t *)malloc(sizeof (nfs_share_list_t));
```

Comment:
I don't see where all these allocations are being freed. Is it expected for the library to leak memory?